Rewrite network again

DaMatrix opened the pull request:
basically i want to keep it as simple to use as it was before, but add the customization that netty allows for

also gonna actually implement raknet this time lol

DaMatrix committed:
nuke network again because i have nothing better to do obviously

DaMatrix committed:
begin adding some interfaces

DaMatrix committed:
continue adding interfaces

man i make great commit messages, don't i

DaMatrix committed:
add interfaces for all endpoint types

DaMatrix committed:
ok i think this is all the API i'm going to need for now

DaMatrix committed:
aaa i'm implementing TCP and need to take a break

DaMatrix committed:
start trying to make a functional TCP implementation

DaMatrix committed:
well, it compiles

DaMatrix committed:
heck ok i have to go

DaMatrix committed:
man this is turning into a mess and i haven't even done anything

DaMatrix committed:
ok i think i came up with an idea for how to implement this best

DaMatrix committed:
i need to make session generics everywhere now >:(

DaMatrix committed:
Merge remote-tracking branch 'origin/development' into dev/rewrite-network-again

# Conflicts:
#	network/src/main/java/net/daporkchop/lib/network/protocol/netty/NettyConnection.java
#	network/src/main/java/net/daporkchop/lib/network/protocol/netty/tcp/WrapperNioServerSocketChannel.java

DaMatrix committed:
ok actually i don't, now let's start making some progress on implementing TCP again

DaMatrix committed:
in theory a functional packet encoding system

DaMatrix committed:
actually test it lol

DaMatrix committed:
use correct nukkitx repo

DaMatrix committed:
try to fix things xd

DaMatrix committed:
i fixed it

DaMatrix committed:
test

DaMatrix committed:
use a builder-based system for creating engines, allowing us to set custom channel options

DaMatrix committed:
fix up datain and dataout, add a way to write raw bytes

DaMatrix committed:
fiddle around, try to make http example work

SupremeMortal commented on build.gradle:
```diff
@@ -74,7 +74,7 @@ allprojects {
         }
         maven {
             name = "NukkitX Snapshots"
-            url = "https://repo.nukkitx.com/maven-snapshots/"
```
This is the correct one for snapshots.
Maybe you're looking for `maven-releases`

DaMatrix replied:
i was having issues with intellij throwing lots of "not found" errors at me and reverting back to the url i'd been using originally fixed them, but i'll change it back now. thanks!

DaMatrix committed:
allow client channel to be initialized properly before returning

DaMatrix committed:
begin splitting everything up into one module per transport

DaMatrix committed:
gtg rip

DaMatrix committed:
make it run when modularized, and allow custom event groups again

DaMatrix committed:
add AbstractUserSession generics basically everywhere

DaMatrix committed:
maybe i'll make my own pipeline

DaMatrix committed:
add a thing with nodes and events and stuff

DaMatrix committed:
just about ready to start implementing this now actually

DaMatrix committed:
implement pipeline stuff, now i have to refactor everything else to use it

DaMatrix committed:
start trying to use it, and i also kinda have to go so committing this now

DaMatrix committed:
Merge branch 'development' into dev/rewrite-network-again

DaMatrix committed:
commit some stuff because i kinda need to

DaMatrix committed:
why do i like deleting stuff so much

DaMatrix committed:
redo pipelines completely because reasons

like c'mon i spent an entire day on that and then deleted the whole thing literally one day later before i even implemented it

DaMatrix committed:
clean up the pipeline stuff a lot, come up with a decent way to actually send things

DaMatrix committed:
ok now i need to think about what to do about encoding/decoding packets to/from binary

DaMatrix committed:
getting closer to having a functional prototype

DaMatrix committed:
i do believe this will work

DaMatrix committed:
temp commit, gtg

DaMatrix committed:
it runs without exceptions now at least

DaMatrix committed:
ssl!

DaMatrix committed:
try to implement minecraft pinger on new pipeline, fail

DaMatrix committed:
i made it work! (needs some optimization i guess, but a lot better already)

DaMatrix committed:
use a composite bytebuf to prevent so much buffer copying

DaMatrix committed:
consider using epoll, and then don't

DaMatrix committed:
start trying to implement SCTP as a (for the most part) copy/paste from TCP with some changes

abstraction coming soon tm

DaMatrix committed:
implement server! along with demo chat thingy

DaMatrix committed:
move examples back lol

DaMatrix committed:
come up with an idea for how to deal with reliabilities (this doesn't compile)

DaMatrix committed:
move the darn examples again

DaMatrix committed:
more abstraction

DaMatrix committed:
start implementing raknet

DaMatrix committed:
basic raknet session

no handlers or stuff though, and i should probably look into using netty concurrent more often :P

DaMatrix committed:
add a better generic parameter finder thing

DaMatrix committed:
start implementing raknet server, ready for testing!

DaMatrix committed:
delete all the old files

actually not quite all of them but same difference

DaMatrix committed:
start making a new API

DaMatrix committed:
gtg rip

DaMatrix committed:
start implementing promises, futures etc

DaMatrix committed:
gtg again

heck

DaMatrix committed:
add some pools lol

DaMatrix committed:
ok commit again i'm going now

DaMatrix committed:
implement a fixed-count thread pool that also doesn't seem to work

DaMatrix committed:
make fixed-count pool work

DaMatrix committed:
make the blocking fixed-count pool work as well

DaMatrix committed:
ok

DaMatrix committed:
there we go, re-implemented all the stuff i did over the weekend and forgot to push

DaMatrix committed:
add a fixed thread count worker group

DaMatrix committed:
add a better way of implementing listeners

DaMatrix committed:
implement a dynamic latch as well

DaMatrix committed:
Merge branch 'dev/concurrent-rewrite' into dev/rewrite-network-again

DaMatrix committed:
oh noes my maven server is broke

DaMatrix committed:
start making a new system with selectors instead of netty because i'm dumb

DaMatrix committed:
continue trying to re-implement TCP

DaMatrix committed:
ok gtg

DaMatrix committed:
channels no longer exist as a class

DaMatrix committed:
start trying to implement client now

DaMatrix committed:
realize that writing data asynchronously is going to be worse than i thought

DaMatrix committed:
why the heck did i think it would be a good idea to write the whole thing myself from scratch?

DaMatrix committed:
delet channels again

DaMatrix committed:
ok, new idea

DaMatrix committed:
framers get their own special package

DaMatrix committed:
delete pipelines for the second time

DaMatrix committed:
i hope i'm not making another mess i'll have to delete again two days from now

DaMatrix committed:
begin refactoring handlers some more

DaMatrix committed:
fix up tcp stuff (including SSL)

DaMatrix committed:
use session factories instead of protocols

DaMatrix committed:
ok, let's start trying to reimplement the mcping example

DaMatrix committed:
nearly ready for first test

DaMatrix committed:
go around fixing minor bugs here and there and everywhere

DaMatrix committed:
make a mess

DaMatrix committed:
mc ping works now

DaMatrix committed:
kinda got http working

DaMatrix committed:
httpget works now with chunked transfer, which is dumb

what moron designed this protocol, seriously? what's the point? why do we need to split a TCP stream into chunks for reading sequential data?

DaMatrix committed:
start adding a better way of sending packets (more descriptive, less work)

DaMatrix committed:
start trying to implement a http server as well

DaMatrix committed:
basic http server works

DaMatrix committed:
move send methods around

DaMatrix committed:
commit some old stuff so i can merge

DaMatrix committed:
Merge remote-tracking branch 'origin/development' into dev/rewrite-network-again

# Conflicts:
#	network/src/main/java/net/daporkchop/lib/network/packet/PacketRegistry.java
#	network/src/main/java/net/daporkchop/lib/network/packet/UserProtocol.java
#	network/src/main/java/net/daporkchop/lib/network/protocol/netty/sctp/SctpChannel.java
#	network/src/main/java/net/daporkchop/lib/network/protocol/netty/sctp/WrapperNioSctpChannel.java
#	network/src/main/java/net/daporkchop/lib/network/protocol/netty/tcp/WrapperNioSocketChannel.java

DaMatrix committed:
make everything compile correctly

DaMatrix committed:
add flush method again

DaMatrix committed:
start implementing a better custom channelgroup thing

DaMatrix committed:
move all kinds of future/promise stuff around

DaMatrix committed:
fix up ssl again

DaMatrix commented:
hey, y'know what? i need this now, and i don't need SCTP or raknet right now, so i'll merge this and implement those separately

DaMatrix committed:
make everything build correctly

DaMatrix committed:
fix primitive jar output
